Supporting async when filtering.

```cpp
    recv(
        filter fil,
        std::set<control_packet_type> types,
        CompletionToken&& token
    ) {
```

I am trying to filter received packages by their type and was wondering why this function doesn't support async as the normal recv function without filtering does support async.

```cpp
    recv(
        CompletionToken&& token
    ) {
```

Unfortunately this doesn't work. I am calling the function like this:

```cpp
co_await  mqtt_client_->recv(async_mqtt::filter::match, {async_mqtt::control_packet_type::connack}, asio::use_awaitable);
```

and these are the compiler errors:

```bash
====================[ Build | mqtt-broadcaster | ci-clang-release ]=============
/usr/bin/cmake --build /home/magni-the-developer/Documents/framework/build/ci-clang-release --target mqtt-broadcaster -j 30
[1/2] Building CXX object exes/mqtt-broadcaster/CMakeFiles/mqtt-broadcaster.dir/src/main.cpp.o
FAILED: exes/mqtt-broadcaster/CMakeFiles/mqtt-broadcaster.dir/src/main.cpp.o 
/usr/bin/clang++-16 -DASYNC_MQTT_USE_LOG -DONLY_C_LOCALE=0 -DUNITS_DOWNCAST_MODE=1 -I/home/magni-the-developer/Documents/framework/exes/mqtt-broadcaster/inc -I/home/magni-the-developer/Documents/framework/libs/progbase/inc/public -I/home/magni-the-developer/Documents/framework/libs/logger/inc/public -I/home/magni-the-developer/Documents/framework/libs/stx/inc/public -I/home/magni-the-developer/Documents/framework/libs/ipc/inc/public -I/home/magni-the-developer/Documents/framework/libs/dbus_util/inc/public -I/home/magni-the-developer/Documents/framework/libs/configure_options/inc/public -I/home/magni-the-developer/Documents/framework/libs/operation_mode/inc/public -I/home/magni-the-developer/Documents/framework/libs/confman/inc/public -isystem /home/magni-the-developer/Documents/framework/vcpkg-sysroot-clang/x64-linux-clang/include -isystem /home/magni-the-developer/Documents/framework/vcpkg-sysroot-clang/x64-linux-clang/lib/pkgconfig/../../include -isystem /home/magni-the-developer/Documents/framework/vcpkg-sysroot-clang/x64-linux-clang/lib/pkgconfig/../../include/libmount -isystem /home/magni-the-developer/Documents/framework/vcpkg-sysroot-clang/x64-linux-clang/lib/pkgconfig/../../include/blkid -stdlib=libc++ -O3 -DNDEBUG -std=gnu++2b -fcolor-diagnostics -fuse-ld=lld -flto=thin -fvisibility=hidden -fvisibility-inlines-hidden -Weverything -Werror -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-pre-c++14-compat -Wno-pre-c++17-compat -Wno-pre-c++20-compat-pedantic -Wno-c++20-compat -Wno-unused-parameter -Wno-padded -Wno-unused-command-line-argument -Wno-missing-braces -MD -MT exes/mqtt-broadcaster/CMakeFiles/mqtt-broadcaster.dir/src/main.cpp.o -MF exes/mqtt-broadcaster/CMakeFiles/mqtt-broadcaster.dir/src/main.cpp.o.d -o exes/mqtt-broadcaster/CMakeFiles/mqtt-broadcaster.dir/src/main.cpp.o -c /home/magni-the-developer/Documents/framework/exes/mqtt-broadcaster/src/main.cpp
In file included from /home/magni-the-developer/Documents/framework/exes/mqtt-broadcaster/src/main.cpp:1:
In file included from /home/magni-the-developer/Documents/framework/vcpkg-sysroot-clang/x64-linux-clang/include/async_mqtt/all.hpp:13:
/home/magni-the-developer/Documents/framework/vcpkg-sysroot-clang/x64-linux-clang/include/async_mqtt/endpoint.hpp:442:25: error: type 'boost::asio::use_awaitable_t<>' does not provide a call operator
                        token(force_move(pv));
                        ^~~~~
/home/magni-the-developer/Documents/framework/exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp:123:33: note: in instantiation of function template specialization 'async_mqtt::basic_endpoint<async_mqtt::role::client, 2, boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>>::recv<const boost::asio::use_awaitable_t<> &>' requested here
        co_await  mqtt_client_->recv(async_mqtt::filter::match, {async_mqtt::control_packet_type::connack}, asio::use_awaitable);
                                ^
/home/magni-the-developer/Documents/framework/exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp:49:44: note: in instantiation of member function 'mqtt_broadcaster<tfc::ipc_ruler::ipc_manager_client &, sdbusplus::bus::match::match, async_mqtt::basic_endpoint<async_mqtt::role::client, 2, boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>>>::connect_to_broker' requested here
    asio::co_spawn(mqtt_client_->strand(), connect_to_broker(), asio::detached);
                                           ^
/home/magni-the-developer/Documents/framework/exes/mqtt-broadcaster/src/main.cpp:32:7: note: in instantiation of member function 'mqtt_broadcaster<tfc::ipc_ruler::ipc_manager_client &, sdbusplus::bus::match::match, async_mqtt::basic_endpoint<async_mqtt::role::client, 2, boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>>>::mqtt_broadcaster' requested here
      application(ctx, mqtt_host, mqtt_port, mqtt_username, mqtt_password, ipc_client, mqtt_client);
      ^
In file included from /home/magni-the-developer/Documents/framework/exes/mqtt-broadcaster/src/main.cpp:1:
In file included from /home/magni-the-developer/Documents/framework/vcpkg-sysroot-clang/x64-linux-clang/include/async_mqtt/all.hpp:13:
/home/magni-the-developer/Documents/framework/vcpkg-sysroot-clang/x64-linux-clang/include/async_mqtt/endpoint.hpp:442:25: error: type 'boost::asio::use_awaitable_t<>' does not provide a call operator
                        token(force_move(pv));
                        ^~~~~
/home/magni-the-developer/Documents/framework/vcpkg-sysroot-clang/x64-linux-clang/include/async_mqtt/endpoint.hpp:403:16: note: in instantiation of function template specialization 'async_mqtt::basic_endpoint<async_mqtt::role::client, 2, boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>>::recv<boost::asio::use_awaitable_t<>>' requested here
        return recv(filter::match, force_move(types), std::forward<CompletionToken>(token));
               ^
/home/magni-the-developer/Documents/framework/vcpkg-sysroot-clang/x64-linux-clang/include/async_mqtt/endpoint.hpp:439:25: note: in instantiation of function template specialization 'async_mqtt::basic_endpoint<async_mqtt::role::client, 2, boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>>::recv<boost::asio::use_awaitable_t<>>' requested here
                        recv(force_move(types), force_move(token));
                        ^
/home/magni-the-developer/Documents/framework/exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp:123:33: note: in instantiation of function template specialization 'async_mqtt::basic_endpoint<async_mqtt::role::client, 2, boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>>::recv<const boost::asio::use_awaitable_t<> &>' requested here
        co_await  mqtt_client_->recv(async_mqtt::filter::match, {async_mqtt::control_packet_type::connack}, asio::use_awaitable);
                                ^
/home/magni-the-developer/Documents/framework/exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp:49:44: note: in instantiation of member function 'mqtt_broadcaster<tfc::ipc_ruler::ipc_manager_client &, sdbusplus::bus::match::match, async_mqtt::basic_endpoint<async_mqtt::role::client, 2, boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>>>::connect_to_broker' requested here
    asio::co_spawn(mqtt_client_->strand(), connect_to_broker(), asio::detached);
                                           ^
/home/magni-the-developer/Documents/framework/exes/mqtt-broadcaster/src/main.cpp:32:7: note: in instantiation of member function 'mqtt_broadcaster<tfc::ipc_ruler::ipc_manager_client &, sdbusplus::bus::match::match, async_mqtt::basic_endpoint<async_mqtt::role::client, 2, boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0>>>>::mqtt_broadcaster' requested here
      application(ctx, mqtt_host, mqtt_port, mqtt_username, mqtt_password, ipc_client, mqtt_client);
      ^
2 errors generated.
ninja: build stopped: subcommand failed.
```

--- a/include/async_mqtt/endpoint.hpp
+++ b/include/async_mqtt/endpoint.hpp
@@ -399,8 +399,20 @@ public:
         ASYNC_MQTT_LOG("mqtt_api", info)
             << ASYNC_MQTT_ADD_VALUE(address, this)
             << "recv";
-
-        return recv(filter::match, force_move(types), std::forward<CompletionToken>(token));
+        BOOST_ASSERT(!recv_processing_);
+        recv_processing_ = true;
+        return
+            as::async_compose<
+                CompletionToken,
+                void(packet_variant_type)
+            >(
+                recv_impl{
+                    *this,
+                    filter::match,
+                    force_move(types)
+                },
+                token
+            );
     }
 
     /**
@@ -423,30 +435,20 @@ public:
         ASYNC_MQTT_LOG("mqtt_api", info)
             << ASYNC_MQTT_ADD_VALUE(address, this)
             << "recv";
-
-        return recv(
-            [
-                this,
-                fil,
-                types = force_move(types),
-                token = std::forward<CompletionToken>(token)
-            ]
-            (packet_variant pv) mutable {
-                if (auto type_opt = pv.type()) {
-                    if ((fil == filter::match  && types.find(*type_opt) == types.end()) ||
-                        (fil == filter::except && types.find(*type_opt) != types.end())
-                    ) {
-                        recv(force_move(types), force_move(token));
-                    }
-                    else {
-                        token(force_move(pv));
-                    }
-                }
-                else {
-                    token(force_move(pv));
-                }
-            }
-        );
+        BOOST_ASSERT(!recv_processing_);
+        recv_processing_ = true;
+        return
+            as::async_compose<
+                CompletionToken,
+                void(packet_variant_type)
+            >(
+                recv_impl{
+                    *this,
+                    fil,
+                    force_move(types)
+                },
+                token
+            );
     }
 
     /**
@@ -1328,6 +1330,8 @@ private: // compose operation impl
 
     struct recv_impl {
         this_type& ep;
+        optional<filter> fil = nullopt;
+        std::set<control_packet_type> types = {};
         optional<system_error> decided_error = nullopt;
         enum { initiate, disconnect, close, complete } state = initiate;
 
@@ -1787,7 +1791,31 @@ private: // compose operation impl
                 );
                 ep.reset_pingreq_recv_timer();
                 ep.recv_processing_ = false;
-                if (call_complete && !decided_error) self.complete(force_move(v));
+
+                if (fil) {
+                    if (auto type_opt = v.type()) {
+                        if ((*fil == filter::match  && types.find(*type_opt) == types.end()) ||
+                            (*fil == filter::except && types.find(*type_opt) != types.end())
+                        ) {
+                            // read the next packet
+                            state = initiate;
+                            auto& a_ep{ep};
+                            as::dispatch(
+                                a_ep.strand(),
+                                force_move(self)
+                            );
+                        }
+                        else {
+                            if (call_complete && !decided_error) self.complete(force_move(v));
+                        }
+                    }
+                    else {
+                        if (call_complete && !decided_error) self.complete(force_move(v));
+                    }
+                }
+                else {
+                    if (call_complete && !decided_error) self.complete(force_move(v));
+                }
             } break;
             case close:
                 BOOST_ASSERT(decided_error);

--- a/test/system/st_keep_alive.cpp
+++ b/test/system/st_keep_alive.cpp
@@ -110,6 +110,8 @@ BOOST_AUTO_TEST_CASE(v5_timeout) {
                 BOOST_TEST(pv->get_if<am::v5::connack_packet>());
                 ep().set_pingreq_send_interval_ms_for_test(10000);
                 yield ep().recv(am::filter::except, {am::control_packet_type::pingresp}, *this);
+                BOOST_TEST(pv->get_if<am::v5::disconnect_packet>());
+                yield ep().recv(am::filter::except, {am::control_packet_type::pingresp}, *this);
                 BOOST_TEST(!*pv); // error as expected
                 set_finish();
             }

--- a/test/unit/CMakeLists.txt
+++ b/test/unit/CMakeLists.txt
@@ -11,6 +11,7 @@ list(APPEND check_PROGRAMS
     ut_ep_basic.cpp
     ut_ep_con_discon.cpp
     ut_ep_topic_alias.cpp
+    ut_ep_recv_filter.cpp
     ut_ep_recv_max.cpp
     ut_ep_size_max.cpp
     ut_ep_store.cpp

--- a/test/unit/ut_ep_recv_filter.cpp
+++ b/test/unit/ut_ep_recv_filter.cpp
@@ -1,0 +1,89 @@
+// Copyright Takatoshi Kondo 2023
+//
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#include "../common/test_main.hpp"
+#include "../common/global_fixture.hpp"
+
+#include <thread>
+
+#include <boost/asio.hpp>
+
+#include <async_mqtt/endpoint.hpp>
+
+#include "stub_socket.hpp"
+#include "packet_compare.hpp"
+
+BOOST_AUTO_TEST_SUITE(ut_ep_recv_filter)
+
+namespace am = async_mqtt;
+namespace as = boost::asio;
+
+// To test the bug fix for https://github.com/redboltz/async_mqtt/issues/42
+BOOST_AUTO_TEST_CASE(recv_filter) {
+    auto version = am::protocol_version::v5;
+    as::io_context ioc;
+    auto guard = as::make_work_guard(ioc.get_executor());
+    std::thread th {
+        [&] {
+            ioc.run();
+        }
+    };
+
+    am::endpoint<async_mqtt::role::client, async_mqtt::stub_socket> ep{
+        version,
+        // for stub_socket args
+        version,
+        ioc
+    };
+
+    auto connect = am::v5::connect_packet{
+        true,   // clean_start
+        0x1234, // keep_alive
+        am::allocate_buffer("cid1"),
+        am::nullopt, // will
+        am::allocate_buffer("user1"),
+        am::allocate_buffer("pass1"),
+        am::properties{}
+    };
+
+    auto connack = am::v5::connack_packet{
+        false,   // session_present
+        am::connect_reason_code::success,
+        am::properties{
+            am::property::receive_maximum{2}
+        }
+    };
+
+    ep.next_layer().set_recv_packets(
+        {
+            // receive packets
+            connack,
+        }
+    );
+
+    // send connect
+    ep.next_layer().set_write_packet_checker(
+        [&](am::packet_variant wp) {
+            BOOST_TEST(am::packet_compare(connect, wp));
+        }
+    );
+    {
+        auto ec = ep.send(connect, as::use_future).get();
+        BOOST_TEST(!ec);
+    }
+
+    // recv connack
+    {
+        auto pv = ep.recv(am::filter::match, {am::control_packet_type::connack}, as::use_future).get();
+        BOOST_TEST(am::packet_compare(connack, pv));
+    }
+
+    ep.close(as::use_future).get();
+    guard.reset();
+    th.join();
+}
+
+BOOST_AUTO_TEST_SUITE_END()

